iceoryx_hoofs/source/posix_wrapper/thread.cpp

```diff
@@ -16,38 +16,97 @@
-        // String length limit is ensured through cxx::string
-        // ERANGE (string too long) intentionally not handled to avoid untestable and dead code
-        std::cerr << "This should never happen! " << r.getHumanReadableErrnum() << std::endl;
-        cxx::Ensures(false && "internal logic error");
-    });
+    if (m_isJoinable)
+    {
+        auto joinResult = posixCall(pthread_join)(m_threadHandle, nullptr).successReturnValue(0).evaluate();
+        if (joinResult.has_error())
+        {
+            switch (joinResult.get_error().errnum)
+            {
+            case EDEADLK:
+                LogError() << "A deadlock was detected when attempting to join the thread.";
+                break;
+            default:
+                LogError() << "This should never happen. An unknown error occurred.";
+                break;
+            }
+        }
+        m_isJoinable = false;
+    }
 }
 
-ThreadName_t getThreadName(pthread_t thread) noexcept
+void Thread::setThreadName(const ThreadName_t& name) noexcept
+{
+    posixCall(iox_pthread_setname_np)(m_threadHandle, name.c_str())
+        .successReturnValue(0)
+        .evaluate()
+        .or_else([](auto& r) {
+            // String length limit is ensured through cxx::string
+            // ERANGE (string too long) intentionally not handled to avoid untestable and dead code
+            LogError() << "This should never happen! " << r.getHumanReadableErrnum();
+            cxx::Ensures(false && "internal logic error");
+            /// @todo thread specific comm file under /proc/self/task/[tid]/comm is read. Opening this file can fail
+            /// and errors possible for open(2) can be retrieved. Handle them here?
+        });
+}
+
+ThreadName_t Thread::getThreadName() noexcept
 {
     char tempName[MAX_THREAD_NAME_LENGTH + 1U];
 
-    posixCall(pthread_getname_np)(thread, tempName, MAX_THREAD_NAME_LENGTH + 1U)
+    posixCall(pthread_getname_np)(m_threadHandle, tempName, MAX_THREAD_NAME_LENGTH + 1U)
         .successReturnValue(0)
         .evaluate()
         .or_else([](auto& r) {
             // String length limit is ensured through MAX_THREAD_NAME_LENGTH
             // ERANGE (string too small) intentionally not handled to avoid untestable and dead code
-            std::cerr << "This should never happen! " << r.getHumanReadableErrnum() << std::endl;
+            LogError() << "This should never happen! " << r.getHumanReadableErrnum();
             cxx::Ensures(false && "internal logic error");
+            /// @todo thread specific comm file under /proc/self/task/[tid]/comm is read. Opening this file can fail
+            /// and errors possible for open(2) can be retrieved. Handle them here?
         });
 
     return ThreadName_t(cxx::TruncateToCapacity, tempName);
 }
 
+ThreadError Thread::errnoToEnum(const int errnoValue) noexcept
+{
+    switch (errnoValue)
+    {
+    case EAGAIN:
+        LogError() << "insufficient resources to create another thread";
```

Comment:
I think it would be nice to add the thread name to every message if a thread name was set up. So something like `getThreadName` and when the string is not empty add something like:

```cxx
auto details = getThreadName();
if ( !details.empty() ) details = ", thread = \"" + details + "\"";
LogError() << "insufficient resources to create another thread" << details;
```

Reply:
We have to be careful with this. If this is called from `getThreadName` at some point, we might have infinite recursion.

It also does not make sense in this case since the only place where this is currently used is in the builder which would give us the wrong thread name.

Reply:
I would like to provide as much information as possible to the user when something goes wrong. 

We could add a comment stating that `getThreadName` never calls this (inside the implementation). When we implement MC/DC tests for this class this case would be covered and so it never should occur in theory.

We could also store the thread name in a separate member variable and since it is quite short it should not blow up the object.